lib/ansible/plugins/connection/ssh.py

```diff
@@ -309,8 +309,49 @@ def _examine_output(self, source, state, b_chunk, sudoable):
 
         return b''.join(output), remainder
 
+
     def _run(self, cmd, in_data, sudoable=True):
         '''
+        Will retry execution if ssh were unable to connect or timeout
+        '''
+        remaining_tries = int(C.ANSIBLE_SSH_RETRIES) + 1
+        cmd_summary = ' '.join(cmd)
+        for attempt in range(remaining_tries):
+            try:
+                return_tuple = self._run_once(cmd=cmd, in_data=in_data, sudoable=sudoable)
+                # 0 = success
+                # 1-254 = remote command return code
+                # 255 = failure from the ssh command itself
+                if return_tuple[0] != 255:
+                    break
+                else:
+                    raise AnsibleConnectionFailure("Failed to connect to the host via ssh.")
+            except (AnsibleConnectionFailure, AnsibleConnectionTimeout, Exception) as e:
```

Comment:
I'm tempted to suggest that the exception could indicate if it was a 'retry-able' exception. Either based on exception type or possibly via an attribute on the exception.

But... I don't see any obvious 'un-retry-able' exceptions. I would argue that authentication related failures shouldn't be retried, but I can see why people may disagree with me (auth failures can be transient as well).